packages/firestore/src/util/proto_byte_string.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PlatformSupport } from '../platform/platform';
+
+/**
+ * Immutable class that represents a "proto" byte string.
+ *
+ * Proto byte strings can either be base64 strings or Uint8Arrays when sent
+ * on the wire, but they are always typed to string. This class abstracts away
+ * this differentiation by holding the proto byte string in a common class that
+ * must be converted into a string before being sent as a proto.
+ */
+export class ByteString {
+  private _binaryString: string;
+
+  private constructor(binaryString: string) {
+    this._binaryString = binaryString;
+  }
+
+  static fromBase64String(base64: string): ByteString {
+    const binaryString = PlatformSupport.getPlatform().atob(base64);
+    return new ByteString(binaryString);
+  }
+
+  static fromUint8Array(array: Uint8Array): ByteString {
+    const binaryString = binaryStringFromUint8Array(array);
+    return new ByteString(binaryString);
+  }
+
+  toBase64(): string {
+    return PlatformSupport.getPlatform().btoa(this._binaryString);
+  }
+
+  toUint8Array(): Uint8Array {
+    const buffer = Uint8ArrayFromBinaryString(this._binaryString);
+    return buffer;
+  }
+
+  approximateByteSize(): number {
+    return this._binaryString.length * 2;
+  }
+
+  isEqual(other: ByteString): boolean {
+    return this._binaryString === other._binaryString;
+  }
+}
+
+/**
+ * Helper function to convert an Uint8array to a binary string.
+ */
+export function binaryStringFromUint8Array(array: Uint8Array): string {
+  // We can't call array.map directly because it expects the return type to
+  // be a Uint8Array, whereas we can convert it to a regular array by invoking
+  // map on the Array prototype.
+  return Array.prototype.map
+    .call(array, (char: number) => {
+      return String.fromCharCode(char);
+    })
+    .join('');
```

Comment:
If you want, this could be:

```return array.reduce((prev, current) => prev + String.fromCharCode(current), "");```

I am not sure whether this is more or less efficient than the current version. The most efficient version is probably:

```
  let result = '';
  for (let i = 0; i < array.length; ++i) {
    result += String.fromCharCode(array[i]);
  }
  return result;
```